### PR TITLE
[91porn] fix title & comment extraction

### LINKED ISSUE #3256: Fix 91porn.com
### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2022.03.08.1**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Indonesia

### Description

Fix 91porn.com since it cant extract title and the fix here https://github.com/ytdl-org/youtube-dl/pull/29876 isn't working. thx

### Verbose log

```shell
E:\scraper and downloader\91porn>youtube-dl -v http://91porn.com/view_video.php?viewkey=051a5a363acfd11e3e69
[debug] Command-line config: ['-v', 'http://91porn.com/view_video.php?viewkey=051a5a363acfd11e3e69']
[debug] User config "C:\Users\X\AppData\Roaming\yt-dlp\config": ['--continue', '--format-sort', 'vbr,abr', '--http-chunk-size', '1M', '--buffer-size', '16K', '--sub-lang', 'english', '--embed-metadata', '--abort-on-unavailable-fragment', '-o', '%(title)s [%(id)s].%(ext)s', '--ffmpeg-location', 'D:\\Program Files\\ffmpeg\\bin\\ffmpeg.exe', '--no-lock']
[debug] Encodings: locale cp1252, fs utf-8, out utf-8, err utf-8, pref cp1252
[debug] ytdl-patched version 2022.03.08.1
[debug] ** The command you are running is not yt-dlp.
[debug] ** Please make bug reports at https://github.com/ytdl-patched/ytdl-patched/issues/new/choose instead.
[debug] Lazy loading extractors is disabled
[debug] Python version 3.9.4 (CPython 64bit) - Windows-10-10.0.19041-SP0
[debug] Checking exe version: "D:\Program Files\ffmpeg\bin\ffmpeg.exe" -bsfs
[debug] Checking exe version: "D:\Program Files\ffmpeg\bin\ffprobe" -bsfs
[debug] exe versions: ffmpeg 2022-03-28-git-5ee198f9aa-essentials_build-www.gyan.dev (setts), ffprobe 2022-03-28-git-5ee198f9aa-essentials_build-www.gyan.dev
[debug] Optional libraries: brotli, certifi, Cryptodome, mutagen, sqlite, websockets
[debug] Proxy map: {}
[debug] [91porn] Extracting URL: http://91porn.com/view_video.php?viewkey=051a5a363acfd11e3e69
[91porn] 051a5a363acfd11e3e69: Downloading webpage
[91porn] [debug] Fetching webpage from http://91porn.com/view_video.php?viewkey=051a5a363acfd11e3e69
ERROR: [91porn] 051a5a363acfd11e3e69: Unable to extract title; please report this issue on  https://github.com/yt-dlp/yt-dlp , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "c:\python\python39\lib\site-packages\yt_dlp\extractor\common.py", line 670, in extract
    ie_result = self._real_extract(url)
  File "c:\python\python39\lib\site-packages\yt_dlp\extractor\porn91.py", line 38, in _real_extract
    title = self._search_regex(
  File "c:\python\python39\lib\site-packages\yt_dlp\extractor\common.py", line 1292, in _search_regex
    raise RegexNotFoundError('Unable to extract %s' % _name)


E:\scraper and downloader\91porn>pause
Press any key to continue . . .
```

### PR DESCRIPTION
**Pull request by pmitchell86**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information.

Fix extraction of Title and Comments fields for the `Porn91` info extractor.

I noticed that https://github.com/ytdl-org/youtube-dl/pull/29876 attempts to do the same, but it's not working and appears abandoned. 

Fixes #3256

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/extractor/porn91.py [hunk 1]
```diff
@@ -33,28 +59,29 @@ def _real_extract(self, url):
...
         video_link_url = self._search_regex(
-            r'<textarea[^>]+id=["\']fm-video_link[^>]+>([^<]+)</textarea>',
+            r'document\.write\(\s*strencode2\s*\((\s*(?:\"[^\"]+\")|(?:\'[^\']+\'))\s*\)\s*\)\s*;?',
             webpage, 'video link')
-        videopage = self._download_webpage(video_link_url, video_id)
 
-        info_dict = self._parse_html5_media_entries(url, videopage, video_id)[0]
+        video_link_url = compat_urllib_parse_unquote(video_link_url)
+        video_link_url = self._search_regex(
+            r"src=\'([^\']+)\'", video_link_url, 'video link')
```
Comment (pukkandan) [inline yt_dlp/extractor/porn91.py:71]:
```suggestion
        video_link_url = self._search_regex(
            r"src=\'([^\']+)\'", urllib.parse.unquote(video_link_url), 'unquoted video link')
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/porn91.py [hunk 2]
```diff
@@ -32,29 +57,27 @@ def _real_extract(self, url):
...
 
         comment_count = int_or_none(self._search_regex(
-            r'留言:\s*</span>\s*(\d+)', webpage, 'comment count', fatal=False))
+            r'留言:\s*<span[^>]*>\s*(\d+)\s*</span>', webpage, 'comment count', fatal=False))
 
-        info_dict.update({
+        return {
             'id': video_id,
+            'url': video_link_url,
+            'ext': 'mp4',
```
Comment (pmitchell86) [inline yt_dlp/extractor/porn91.py:78]: This is incorrect as well. There are some examples of `video_link_url` with `m3u8` extension. I'm looking at `_parse_html5_media_entries` to fill in the value unless there is something easier.

Comment (pukkandan) [reply yt_dlp/extractor/porn91.py:78]: https://github.com/yt-dlp/yt-dlp/blob/990dd7b00fc5f7c22ff339d0ca5b5b4f21923dca/yt_dlp/extractor/airtv.py#L63-L69 may be helpful

Comment (pmitchell86) [reply yt_dlp/extractor/porn91.py]: Perfect, thanks!

File: yt_dlp/extractor/porn91.py [hunk 3]
```diff
@@ -32,29 +68,40 @@ def _real_extract(self, url):
...
-            r'<textarea[^>]+id=["\']fm-video_link[^>]+>([^<]+)</textarea>',
+            r'document\.write\(\s*strencode2\s*\(\s*((?:"[^"]+\")|(?:\'[^\']+\'))\s*\)\s*\)',
             webpage, 'video link')
-        videopage = self._download_webpage(video_link_url, video_id)
+        video_link_url = self._search_regex(
+            r"src=\'([^\']+)\'", urllib.parse.unquote(video_link_url), 'unquoted video link')
+
+        release_date = self._search_regex(
+            r'<span\s+class=["\']title-yakov["\']>(\d{4}-\d{2}-\d{2})</span>', webpage, 'timestamp', fatal=False)
+        release_date = release_date.replace('-', '') if release_date else None
```
Comment (pukkandan) [inline yt_dlp/extractor/porn91.py:82]: `unified_strdate`

File: yt_dlp/extractor/porn91.py [hunk 4]
```diff
@@ -32,29 +68,40 @@ def _real_extract(self, url):
...
 
-        info_dict = self._parse_html5_media_entries(url, videopage, video_id)[0]
+        description = self._search_regex(
+            r'<span\s+class=["\']more title["\']>([^<]*)<br', webpage, 'description', fatal=False).strip()
```
Comment (pukkandan) [inline yt_dlp/extractor/porn91.py:85]:
```suggestion
        description = self._html_search_regex(
            r'<span\s+class=["\']more title["\']>([^<]+)<br', webpage, 'description', fatal=False)
```
You realize the .strip negates the whole point of fatal=False, right?

File: yt_dlp/extractor/porn91.py [hunk 5]
```diff
@@ -32,29 +68,40 @@ def _real_extract(self, url):
...
+        title = self._html_extract_title(webpage)
+        title = remove_end(title.replace('\n', ''), 'Chinese homemade video').strip()
 
         video_link_url = self._search_regex(
-            r'<textarea[^>]+id=["\']fm-video_link[^>]+>([^<]+)</textarea>',
+            r'document\.write\(\s*strencode2\s*\(\s*((?:"[^"]+\")|(?:\'[^\']+\'))\s*\)\s*\)',
             webpage, 'video link')
-        videopage = self._download_webpage(video_link_url, video_id)
+        video_link_url = self._search_regex(
+            r"src=\'([^\']+)\'", urllib.parse.unquote(video_link_url), 'unquoted video link')
```
Comment (pukkandan) [inline yt_dlp/extractor/porn91.py:78]:
```suggestion
            r'src=["\']([^\'"]+)["\']', urllib.parse.unquote(video_link_url), 'unquoted video link')
```

File: yt_dlp/extractor/porn91.py [hunk 6]
```diff
@@ -32,29 +68,40 @@ def _real_extract(self, url):
...
+        view_count = int_or_none(self._search_regex(
+            r'热度:\s*<span[^>]*>\s*(\d+)\s*</span>', webpage, 'comment count', fatal=False))
 
-        info_dict.update({
+        return {
             'id': video_id,
+            'url': video_link_url,
+            'ext': determine_ext(video_link_url),
             'title': title,
+            'release_date': release_date,
```
Comment (pukkandan) [inline yt_dlp/extractor/porn91.py:101]: You sure this shouldn't be `upload_date`?

**Review by pukkandan — CHANGES_REQUESTED** [review]
Can you give me an example for m3u8?

File: yt_dlp/extractor/porn91.py (hunk 4 above)
Comment (pmitchell86) [reply yt_dlp/extractor/porn91.py:85]: Oops

**Comment by pmitchell86** [comment]
@pukkandan as for that m3u8 example: I don't have one. The latest version of the PR uses `determine_ext()` which seems to return the correct info in my testing. The files always come down as mp4

File: yt_dlp/extractor/porn91.py [hunk 7]
```diff
@@ -1,26 +1,48 @@
...
+            'view_count': int,
+            'age_limit': 18,
+        }
+    }, {
+        'url': 'https://91porn.com/view_video.php?viewkey=726186267387ffe1e5e6',
+        'md5': 'b0067d158495566fe3d20bafd89e36d1',
+        'info_dict': {
+            'id': '726186267387ffe1e5e6',
+            'title': '见过卖老婆的，那你见过卖亲闺女的吗？',
+            'description': '疫情当下，如何约炮？\n--19kn.cc--\n拥有全国线下学生、少妇、反差婊、兼职良家。\n并且免费！！！\n只需要一个电话，一个定位，就能送炮上门。可提前查看照片\n（妹子自带48小时核酸报告）\n约炮，我们是认真的！\n并且拥有三大优势！\n\n1、各种求包养母狗，学生妹资源。为你解决各种需要。--19kn.cc--\n\n2、所有女性会员经过实名视频验证，平台严选，杜绝各种骗红包，口嗨者。--19kn.cc--\n\n3、5年大平台，91许多约炮案例，包括知名博主女伴，均是我们撮合成功的，保障会员隐私，并且约炮3次可自行联系平台进行信息发布。--19kn.cc--\n\n平台5周年庆活动，特回馈91狼友\n\n1、所有女性会员，如果参假，举报客服，核实成功奖励10000人民币。\n\n2、约炮成功并且反馈客服，赠送91vip自拍达人号\n\n3、情侣入驻，可享受专属奖励（奖金5000元）\n\n年关将近，平台大放血，只为各位狼友能找到固定性伴侣，度过美好新年！\n\n约炮渠道请登录--19kn.cc--\n\nPS:招网络客服，对接客户，安排妹子（要求耐心，熟悉客服流程优先，有电脑优先）工作时间:12小时制，\n\n招男模，女模（要求形象气质佳，需提供体检报告）\n有意可以联系官方招聘邮箱[email\xa0protected]',
```
Comment (pukkandan) [inline yt_dlp/extractor/porn91.py:37]: use md5

File: yt_dlp/extractor/porn91.py [hunk 8]
```diff
@@ -1,26 +1,48 @@
...
 
-    _TEST = {
+    _TESTS = [{
         'url': 'http://91porn.com/view_video.php?viewkey=7e42283b4f5ab36da134',
-        'md5': '7fcdb5349354f40d41689bd0fa8db05a',
+        'md5': 'd869db281402e0ef4ddef3c38b866f86',
         'info_dict': {
             'id': '7e42283b4f5ab36da134',
             'title': '18岁大一漂亮学妹，水嫩性感，再爽一次！',
+            'description': '想看我拍新的系列都请帮我加精跟5星好评哦！希望大家鼎力支持，谢过了。我再重申，这次是朋友介绍安排的漂亮学生，费用不低，不过胜在年轻听话，水嫩性感，很超值的女生（6分05有91验证）。PS:本人强壮耐久，事业型男，愿意结交江浙沪的漂亮学妹，加Q：2889560495，语音验证性别，欢迎女生约我，或者靠谱男来一起泡美眉。',
```
Comment (pukkandan) [inline yt_dlp/extractor/porn91.py:23]: same

File: yt_dlp/extractor/porn91.py [hunk 9]
```diff
@@ -29,32 +51,42 @@ def _real_extract(self, url):
         webpage = self._download_webpage(
             'http://91porn.com/view_video.php?viewkey=%s' % video_id, video_id)
 
-        if '作为游客，你每天只可观看10个视频' in webpage:
-            raise ExtractorError('91 Porn says: Daily limit 10 videos exceeded', expected=True)
+        if '作为游客，你每天只可观看15个视频' in webpage:
```
Comment (pukkandan) [inline yt_dlp/extractor/porn91.py:54]: Why not use regex to extract out the number?

File: yt_dlp/extractor/porn91.py [hunk 10]
```diff
@@ -29,32 +51,42 @@ def _real_extract(self, url):
...
-            r'<div id="viewvideo-title">([^<]+)</div>', webpage, 'title')
-        title = title.replace('\n', '')
+        title = self._html_extract_title(webpage)
+        title = remove_end(title.replace('\n', ''), 'Chinese homemade video').strip()
 
         video_link_url = self._search_regex(
-            r'<textarea[^>]+id=["\']fm-video_link[^>]+>([^<]+)</textarea>',
-            webpage, 'video link')
-        videopage = self._download_webpage(video_link_url, video_id)
+            r'document\.write\(\s*strencode2\s*\(\s*((?:"[^"]+")|(?:\'[^\']+\'))\s*\)\s*\)', webpage, 'video link')
```
Comment (pukkandan) [inline yt_dlp/extractor/porn91.py:61]:
```suggestion
            r'document\.write\(\s*strencode2\s*\(\s*((?:"[^"]+")|(?:\'[^\']+\'))', webpage, 'video link')
```

File: yt_dlp/extractor/porn91.py [hunk 11]
```diff
@@ -29,32 +51,42 @@ def _real_extract(self, url):
...
+        view_count = int_or_none(self._search_regex(
+            r'热度:\s*<span[^>]*>\s*(\d+)\s*</span>', webpage, 'view count', fatal=False))
 
-        info_dict.update({
+        return {
             'id': video_id,
+            'url': video_link_url,
+            'ext': determine_ext(video_link_url),
             'title': title,
+            'upload_date': upload_date,
```
Comment (pukkandan) [inline yt_dlp/extractor/porn91.py:86]:
```suggestion
            'upload_date': unified_strdate(self._search_regex(
                r'<span\s+class=["\']title-yakov["\']>(\d{4}-\d{2}-\d{2})</span>',
                webpage, 'upload_date', fatal=False)),
```
etc

File: yt_dlp/extractor/porn91.py [hunk 12]
```diff
@@ -29,32 +51,42 @@ def _real_extract(self, url):
...
+        upload_date = self._search_regex(
+            r'<span\s+class=["\']title-yakov["\']>(\d{4}-\d{2}-\d{2})</span>', webpage, 'upload_date', fatal=False)
+        upload_date = unified_strdate(upload_date)
+
+        description = self._html_search_regex(
+            r'<span\s+class=["\']more title["\']>\s*(.*(?!</span>))\s*</span>', webpage, 'description', fatal=False)
 
         duration = parse_duration(self._search_regex(
-            r'时长:\s*</span>\s*(\d+:\d+)', webpage, 'duration', fatal=False))
+            r'时长:\s*<span[^>]*>\s*(\d+:\d+:\d+)\s*</span>', webpage, 'duration', fatal=False))
```
Comment (pukkandan) [inline yt_dlp/extractor/porn91.py:73]:
```suggestion
            r'时长:\s*<span[^>]*>\s*(\d+(?::\d+){1,2})', webpage, 'duration', fatal=False))
```
* `{1,2}` to support old format too
* Is `</span>` needed?

File: yt_dlp/extractor/porn91.py [hunk 13]
```diff
@@ -29,32 +51,42 @@ def _real_extract(self, url):
...
+        video_link_url = self._search_regex(
+            r'src=["\']([^"\']+)["\']', urllib.parse.unquote(video_link_url), 'unquoted video link')
 
-        info_dict = self._parse_html5_media_entries(url, videopage, video_id)[0]
+        upload_date = self._search_regex(
+            r'<span\s+class=["\']title-yakov["\']>(\d{4}-\d{2}-\d{2})</span>', webpage, 'upload_date', fatal=False)
+        upload_date = unified_strdate(upload_date)
+
+        description = self._html_search_regex(
+            r'<span\s+class=["\']more title["\']>\s*(.*(?!</span>))\s*</span>', webpage, 'description', fatal=False)
```
Comment (pukkandan) [inline yt_dlp/extractor/porn91.py:70]: The regex is wrong. You have to group the `.` with `(?!)` and `*` the whole thing to do what you want. But it's better to just do:
```suggestion
            r'<span\s+class=["\']more title["\']>\s*([^<]+', webpage, 'description', fatal=False)
```

File: yt_dlp/extractor/porn91.py [hunk 14]
```diff
@@ -29,32 +51,42 @@ def _real_extract(self, url):
...
+            r'留言:\s*<span[^>]*>\s*(\d+)\s*</span>', webpage, 'comment count', fatal=False))
+
+        view_count = int_or_none(self._search_regex(
+            r'热度:\s*<span[^>]*>\s*(\d+)\s*</span>', webpage, 'view count', fatal=False))
 
-        info_dict.update({
+        return {
             'id': video_id,
+            'url': video_link_url,
+            'ext': determine_ext(video_link_url),
```
Comment (pukkandan) [inline yt_dlp/extractor/porn91.py:84]:
```suggestion
```
Unnecessary

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/porn91.py [hunk 15]
```diff
@@ -1,26 +1,48 @@
...
+            'age_limit': 18,
+        }
+    }, {
+        'url': 'https://91porn.com/view_video.php?viewkey=726186267387ffe1e5e6',
+        'md5': 'b0067d158495566fe3d20bafd89e36d1',
+        'info_dict': {
+            'id': '726186267387ffe1e5e6',
+            'title': '见过卖老婆的，那你见过卖亲闺女的吗？',
+            'description': '疫情当下，如何约炮？\n--19kn.cc--\n拥有全国线下学生、少妇、反差婊、兼职良家。\n并且免费！！！\n只需要一个电话，一个定位，就能送炮上门。可提前查看照片\n（妹子自带48小时核酸报告）\n约炮，我们是认真的！\n并且拥有三大优势！\n\n1、各种求包养母狗，学生妹资源。为你解决各种需要。--19kn.cc--\n\n2、所有女性会员经过实名视频验证，平台严选，杜绝各种骗红包，口嗨者。--19kn.cc--\n\n3、5年大平台，91许多约炮案例，包括知名博主女伴，均是我们撮合成功的，保障会员隐私，并且约炮3次可自行联系平台进行信息发布。--19kn.cc--\n\n平台5周年庆活动，特回馈91狼友\n\n1、所有女性会员，如果参假，举报客服，核实成功奖励10000人民币。\n\n2、约炮成功并且反馈客服，赠送91vip自拍达人号\n\n3、情侣入驻，可享受专属奖励（奖金5000元）\n\n年关将近，平台大放血，只为各位狼友能找到固定性伴侣，度过美好新年！\n\n约炮渠道请登录--19kn.cc--\n\nPS:招网络客服，对接客户，安排妹子（要求耐心，熟悉客服流程优先，有电脑优先）工作时间:12小时制，\n\n招男模，女模（要求形象气质佳，需提供体检报告）\n有意可以联系官方招聘邮箱[email\xa0protected]',
+            'ext': 'm3u8',
```
Comment (pukkandan) [inline yt_dlp/extractor/porn91.py:38]: Wrong. Either test def is wrong, or you need to use `_extract_m3u8_formats_and_subtitles` in code

File: yt_dlp/extractor/porn91.py [hunk 16]
```diff
@@ -1,26 +1,48 @@
...
+    remove_end,
+    unified_strdate,
     ExtractorError,
 )
 
 
 class Porn91IE(InfoExtractor):
     IE_NAME = '91porn'
-    _VALID_URL = r'(?:https?://)(?:www\.|)91porn\.com/.+?\?viewkey=(?P<id>[\w\d]+)'
+    _VALID_URL = r'(?:https?://)(?:www\.|)91porn\.com/.*([\?&])viewkey=(?P<id>[\w\d]+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/porn91.py:15]:
```suggestion
    _VALID_URL = r'(?:https?://)(?:www\.|)91porn\.com/view_video.php\?([^#]+&)?viewkey=(?P<id>\w+)'
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by pukkandan** [comment]
Explain https://github.com/yt-dlp/yt-dlp/pull/5932/commits/6a7a5518e466d3e73feccab51138929154450c38. You shouldn't just pass master m3u8 without processing

**Comment by pmitchell86** [comment]
> Explain [6a7a551](https://github.com/yt-dlp/yt-dlp/commit/6a7a5518e466d3e73feccab51138929154450c38). You shouldn't just pass master m3u8 without processing

In 6a7a551 all the urls i used as examples would download as `*.m3u8` and not be playable. Renaming the file to mp4 enabled VLC to play the files.

**Comment by pukkandan** [comment]
If it's an m3u8, use `_extract_m3u8_formats_and_subtitles`. Like I said before,

> You shouldn't just pass master m3u8 without processing

**Comment by pmitchell86** [comment]
`https://91porn.com/view_video.php?viewkey=7ef0cf3d362c699ab91c` is an example (taken from the tests) where 
- `determine_ext()` returns `m3u8` 
-`_extract_m3u8_formats_and_subtitles()` returns:  `[{'format_id': '', 'format_index': None, 'url': 'https://cdn77.91p49.com/m3u8/731633/731633.m3u8', 'ext': None, 'protocol': 'm3u8_native', 'preference': None, 'quality': None, 'has_drm': None}]`

if the hard-coded `'ext': 'mp4',` is removed from the return dict then the downloaded file will have an `.m3u8` extension.

File: yt_dlp/extractor/porn91.py [hunk 17]
```diff
@@ -29,32 +51,47 @@ def _real_extract(self, url):
...
+                r'留言:\s*<span[^>]*>\s*(\d+)\s*</span>', webpage, 'comment count', fatal=False)),
+            'view_count': int_or_none(self._search_regex(
+                r'热度:\s*<span[^>]*>\s*(\d+)\s*</span>', webpage, 'view count', fatal=False)),
+            'age_limit': 18,
+        }
+
+    def _get_formats_and_subtitle(self, video_link_url, video_id):
+        ext = determine_ext(video_link_url)
+        if ext == 'm3u8':
+            formats, subtitles = self._extract_m3u8_formats_and_subtitles(video_link_url, video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/porn91.py:92]:
```suggestion
            formats, subtitles = self._extract_m3u8_formats_and_subtitles(video_link_url, video_id, ext='mp4')
```
This should fix any issues you are having with ext

File: yt_dlp/extractor/porn91.py [hunk 18]
```diff
@@ -29,32 +51,47 @@ def _real_extract(self, url):
...
 
-        info_dict.update({
+        return {
             'id': video_id,
-            'title': title,
-            'duration': duration,
-            'comment_count': comment_count,
-            'age_limit': self._rta_search(webpage),
-        })
+            'url': video_link_url,
```
Comment (pukkandan) [inline yt_dlp/extractor/porn91.py:71]:
```suggestion
```

File: yt_dlp/extractor/porn91.py [hunk 19]
```diff
@@ -29,32 +51,47 @@ def _real_extract(self, url):
...
+            'title': remove_end(self._html_extract_title(webpage).replace('\n', ''), 'Chinese homemade video').strip(),
+            'formats': formats,
+            'subtitles': subtitles,
+            # 'ext': 'mp4', # Include this so files are named *.mp4 instead of *.m3u8
```
Comment (pukkandan) [inline yt_dlp/extractor/porn91.py:75]:
```suggestion
```

**Review by pukkandan — APPROVED** [review]

File: yt_dlp/extractor/porn91.py (hunk 17 above)
Comment (pmitchell86) [reply yt_dlp/extractor/porn91.py]: That fixed it!

**Pull request merged**